cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ETLStage.java

```diff
@@ -141,18 +144,23 @@ public ETLStage updateStage(ApplicationUpdateContext updateContext) throws Excep
    * @return Updated plugin object to be used for the udated stage. Returned null if no changes to current plugin.
    */
   private ETLPlugin upgradePlugin(ApplicationUpdateContext updateContext) throws Exception {
-    // Currently tries to find latest plugin in SYSTEM scope and upgrades current plugin if version is higher,
-    // ignoring current plugin scope.
-    // In future, we can modify logic to fetch the latest plugin in any scope.
-    List<ArtifactId> candidates =
-      updateContext.getPluginArtifacts(plugin.getType(), plugin.getName(), ArtifactScope.SYSTEM, null);
-    if (candidates.isEmpty()) {
+
+    // Find candidate plugins from all artifact scopes to consider for this upgrade operation.
+    List<ArtifactId> candidates = new ArrayList<>();
+    for (ArtifactScope scope: updateContext.getAllowedArtifactScopes()) {
```

Comment:
I don't think it is up to the app to decide what scope to include since the choice is made by the user who call the REST api.